Substrait string concat with more than two arguments fails conversion to Calcite

The Substrait `concat` function is variadic so allows more than two arguments. When converting a Substrait concat expression with three (or more) arguments to Calcite, an error occurs since Calcite's corresponding `||` operator only allows two operands:

```
java.lang.AssertionError: wrong operand count 3 for ||
	at org.apache.calcite.util.Litmus.lambda$static$0(Litmus.java:31)
	at org.apache.calcite.sql.SqlBinaryOperator.validRexOperands(SqlBinaryOperator.java:229)
	at org.apache.calcite.rex.RexCall.<init>(RexCall.java:93)
	at org.apache.calcite.rex.RexBuilder.makeCall(RexBuilder.java:293)
	at org.apache.calcite.rex.RexBuilder.makeCall(RexBuilder.java:279)
	at io.substrait.isthmus.expression.ExpressionRexConverter.visit(ExpressionRexConverter.java:433)
```

This could be resolved by modifying the expression during conversion to be nested / chained expressions, each with only two arguments. For example, `concat(a, b, c)` becomes `concat(concat(a, b), c)`. This results in the expected Calcite (and SQL) output.

--- a/isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java
+++ b/isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java
@@ -521,6 +521,13 @@ public class ExpressionRexConverter
             .collect(Collectors.toList());
 
     RelDataType returnType = typeConverter.toCalcite(typeFactory, expr.outputType());
+    if (operator == SqlStdOperatorTable.CONCAT && args.size() > 2) {
+      return args.stream()
+          .skip(1)
+          .reduce(
+              args.get(0),
+              (left, right) -> rexBuilder.makeCall(returnType, operator, List.of(left, right)));
+    }
     return rexBuilder.makeCall(returnType, operator, args);
   }
 

--- a/isthmus/src/test/java/io/substrait/isthmus/FunctionConversionTest.java
+++ b/isthmus/src/test/java/io/substrait/isthmus/FunctionConversionTest.java
@@ -20,6 +20,7 @@ import java.util.stream.Stream;
 import org.apache.calcite.rex.RexCall;
 import org.apache.calcite.rex.RexNode;
 import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
 import org.junit.jupiter.api.Test;
 
 /**
@@ -210,6 +211,31 @@ class FunctionConversionTest extends PlanTestBase {
   @Test
   void concatStringLiteralAndChar() throws Exception {
     assertProtoPlanRoundrip("select 'brand_'||P_BRAND from PART");
+  }
+
+  @Test
+  void variadicConcat() {
+    ScalarFunctionInvocation concat =
+        sb.scalarFn(
+            DefaultExtensionCatalog.FUNCTIONS_STRING,
+            "concat:str",
+            TypeCreator.REQUIRED.STRING,
+            Expression.StrLiteral.builder().value("a").build(),
+            Expression.StrLiteral.builder().value("b").build(),
+            Expression.StrLiteral.builder().value("c").build());
+
+    RexCall outer =
+        assertInstanceOf(
+            RexCall.class, concat.accept(expressionRexConverter, Context.newContext()));
+    assertEquals(SqlStdOperatorTable.CONCAT, outer.getOperator());
+    assertEquals(2, outer.getOperands().size());
+
+    RexCall inner = assertInstanceOf(RexCall.class, outer.getOperands().get(0));
+    assertEquals(SqlStdOperatorTable.CONCAT, inner.getOperator());
+    assertEquals(2, inner.getOperands().size());
+    assertEquals("'a':VARCHAR", inner.getOperands().get(0).toString());
+    assertEquals("'b':VARCHAR", inner.getOperands().get(1).toString());
+    assertEquals("'c':VARCHAR", outer.getOperands().get(1).toString());
   }
 
   @Test

